### PR TITLE
M2P-372 Bugfix: Total mismatch with Mirasvit rewards

### PR DESCRIPTION
**Pull request by dumega**
# Description
Mirasvit rewards supports two different points spending styles Flexible (partial_point) and Fixed (full_point), and these two rules have different rounding logic on the available points number. 
For Flexible, it allows to apply float discount amount.
For Fixed, it only allows to apply integer discount amount.

This PR is to support these rules.

Fixes: https://boltpay.atlassian.net/browse/M2P-372

#changelog M2P-372 Bugfix: Total mismatch with Mirasvit rewards

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Review by vitaliyreznikov — APPROVED** [review]

File: ThirdPartyModules/Mirasvit/Rewards.php [hunk 1]
```diff
@@ -268,19 +268,52 @@ private function getMirasvitRewardsAmount($quote)
                 $websiteId = $quote->getStore()->getWebsiteId();
                 $rules = $this->mirasvitRewardsSpendRulesListHelper->getRuleCollection($websiteId, $customer->getGroupId());
                 if ($rules->count()) {
+                    $totalAmount = 0;
+                    $totalPoints = 0;
+        
+                    $mockSubtotal = 999999999;
```
Comment (johnmccombs1) [inline ThirdPartyModules/Mirasvit/Rewards.php:274]: Can you explain why set the value to 999999999 here?

Comment (dumega) [reply ThirdPartyModules/Mirasvit/Rewards.php:274]: @johnmccombs1 Cause we always try to apply full available rewards amount to Bolt cart, we create a mock subtotal large enough to implement this purpose.

**Comment by dumega** [comment]
@vitaliyreznikov @johnmccombs1 

Update: I add support for Mirasvit_Rewards multiple spending rules

**Review by danac-gs — APPROVED** [review]
LGTM (although a bit complicated, but if it works it's OK)

File: ThirdPartyModules/Mirasvit/Rewards.php (hunk 1 above)
Comment (vitaliyreznikov) [reply ThirdPartyModules/Mirasvit/Rewards.php:296]: @dumega Can we create a class constant instead of a variable for this number?

**Review by vitaliyreznikov — COMMENTED** [review]
This approach looks so complicated. Do I understand correctly you copied some code from the 3p module?
Any chance we can call this code instead of copying?

**Comment by dumega** [comment]
@vitaliyreznikov `function isRuleValid` and `function calcPointsPerRule` are both private functions in the 3rd-party module, and `function calMaxSpendAmount`. `function getCartPointsRange` and `function calcPointsPerRule` all have our own custom logic, so we can not directly call them from the 3rd-party module

File: ThirdPartyModules/Mirasvit/Rewards.php (hunk 1 above)
Comment (dumega) [reply ThirdPartyModules/Mirasvit/Rewards.php]: @vitaliyreznikov Done

**Comment by vitaliyreznikov** [comment]
@dumega Got it! Could you please add comments about this into each phpdoc block? Like this function is copied from class ... because it is private in the class.

**Review by vitaliyreznikov — COMMENTED** [review]
@dumega Got it! Could you please add comments about this into each phpdoc block? Like this function is copied from class ... because it is private in the class.

**Comment by dumega** [comment]
@vitaliyreznikov  Done

**Review by danac-gs — APPROVED** [review]
LGTM. One note: could we use reflection or closures for calling private methods instead of copying them?

**Review by vitaliyreznikov — APPROVED** [review]

**Review by johnmccombs1 — APPROVED** [review]

**Pull request merged**